Changes to support DLT

pj-spoelders opened the pull request:
Closes #671 

# Checklist

Note: If any of the items in the checklist are not relevant to your PR, leave the box unchecked.

## For any Pull Request

Is the following correct:
* [x] the title of the Pull Request?
* [x] the title of the corresponding issue?
* [x] there are no other open [Pull Requests](../../../../pulls) for the same update/change? <!-- markdown-link-check-disable-line -->
* [x] that the issue which this Pull Request fixes ("Fixes...") is mentioned?

## When Changes Were Made

Did you:
* [ ] update the changelog?
* [x] update the implementation?
* [x] check coverage and add tests: unit tests and, if relevant, integration tests?
* [ ] update the User Guide & other documentation?
* [ ] resolve any failing CI criteria (incl. Sonar quality gate)?

## When Preparing a Release

Have you:
* [ ] thought about version number (major, minor, patch)?
* [ ] checked Exasol packages for updates and resolved open vulnerabilities, if easily possible?

pj-spoelders committed:
Enable dlt to build metadata schema and tables

pj-spoelders committed:
Fix for inserting timestamps

pj-spoelders committed:
Add an exception handler for PyExasol errors bubbling up and convert them to proper PEP-249 exceptions

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -692,8 +693,52 @@ def visit_big_integer(self, type_, **kw):
+    # --- Date/time ---
+
+    # Some SQLAlchemy versions dispatch DateTime() to 'DATETIME' (upper-case)
+    def visit_DATETIME(self, type_, **kw):
```
Is it intentional that they aren't built referring to one another? For instance, we could have visit_DATETIME as it currently is defined and then have visit_datetime call self.visit_DATETIME and pass on the values.

pj-spoelders replied:
Fully agree that this needs to be cleaned up a bit and we reuse methods if they're available/work in our case.

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -692,8 +693,52 @@ def visit_big_integer(self, type_, **kw):
+    # --- Strings / Text ---
+
+     # SA String(length) -> VARCHAR(n); String() -> CLOB (Exasol requires length)
+    def visit_string(self, type_, **kw):
```
Similar question. Why not call `visit_VARCHAR` or `visit_CLOB`?

pj-spoelders replied:
Fully agree that this needs to be cleaned up a bit and we reuse methods if they're available/work in our case.

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -692,8 +693,52 @@ def visit_big_integer(self, type_, **kw):
+        return "DECIMAL(18,0)"
+
+    # Some code paths use DECIMAL directly rather than numeric
+    def visit_DECIMAL(self, type_, **kw):
```
Just curious as to how these were arrived to: Were the other `visit_*` functions checked or were these ones that caused difficulties for Exasol with dlt testing?

pj-spoelders replied:
Yes this is just enough to get DLT working right now, there's probably some/plenty of other cases that might still not work.

pj-spoelders replied:
I've added support for time and exceptions for unsupported binary types.

pj-spoelders replied:
Unless I missed something obvious I think this covers most types ..

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -1199,3 +1279,33 @@ def fkey_rec():
+        try:
+            return super().do_execute(cursor, statement, parameters, context)
+        except Exception as e:
+            mapped = _PYEXA_TO_SA.get(e.__class__.__name__)
```
We might be able to simplify the exceptions here, as the Python Exception case allows for more than 1 to be checked, and we could extract that to a function if other usages needed to be covered.

However, it's possible you have something more in mind, as it looks like `do_execute` maybe doesn't cover all execution points.

pj-spoelders replied:
Let's have a look at how you'd do this this afternoon ..

pj-spoelders replied:
Rewrote this. Hope it is as expected.

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -1199,3 +1279,33 @@ def fkey_rec():
         """EXASolution has no explicit indexes"""
         return []
+
+    def type_descriptor(self, typeobj):
```
Did you try putting the ExaTimestamp in the `colspecs` mapping?

https://github.com/exasol/sqlalchemy-exasol/blob/master/sqlalchemy_exasol/websocket.py#L91

pj-spoelders replied:
Let's try this out.

pj-spoelders replied:
I've tried this today but I ran into a plethora of issues, one after the other: Circular import/dependency which then needs to be solved with lazy population of colspecs, etc ...
I think the type_descriptor approach here is cleaner. If you vehemently disagree, do let me know, but otherwise I'd leave it like this.

ArBridgeman replied:
Thanks for trying it. No, it's ok for you to leave it like this.

pj-spoelders committed:
Merge branch 'master' into feature/add_dlt_support

pj-spoelders committed:
Reworked exception mapping/translation (PR feedback)

pj-spoelders commented on sqlalchemy_exasol/base.py:
```diff
@@ -692,8 +692,52 @@ def visit_big_integer(self, type_, **kw):
+    # --- Date/time ---
+
+    # Some SQLAlchemy versions dispatch DateTime() to 'DATETIME' (upper-case)
+    def visit_DATETIME(self, type_, **kw):
```
So I checked this a bit more in depth: 
Just calling the other 'visit_xxx' function is messy as there are type hints for the '_type' arguments of these methods and they differ per method.
Cleanest I can get it is using some helper methods instead of returning a magic string I think?

ArBridgeman replied:
Ok, that sounds alright.

pj-spoelders committed:
expanded EXATypeCompiler
- support more types
- removed old ODBC overrides

pj-spoelders committed:
Unit tests for exacompiler, exception mapper, and timestamp handling via type descriptor

pj-spoelders committed:
formatting fixes

pj-spoelders committed:
Merge branch 'master' into feature/add_dlt_support

pj-spoelders committed:
Map time to varchar(16)

pj-spoelders committed:
Throw exceptions for BLOB/BINARY/VARBINARY types (not supported)

pj-spoelders committed:
Create table and insert statement test for compiler, type descriptor

pj-spoelders committed:
fix formatting

pj-spoelders committed:
fixes and changes to tests pertaining to binary types

pj-spoelders committed:
Fix formatting

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -686,22 +697,107 @@ def visit_drop_index(self, drop, **kw):
-    dialect. However, BIGINT is mapped to DECIMAL(36) and the driver only
-    converts types decimal scale==0 and 9<precision<=19 back to BIGINT
-    https://www.exasol.com/support/browse/EXA-23267
+    """Type compiler for Exasol (WebSocket-only).
```
Maybe instead of putting Websocket-only, as this base.py class would be used for other dialects, you can put say that it's been only tested for that or something similar.

pj-spoelders replied:
Fair enough!

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -686,22 +697,107 @@ def visit_drop_index(self, drop, **kw):
+    Policy:
+    - Do NOT override types Exasol already supports well via native names/aliases
+      (BIGINT/INTEGER/SMALLINT, NUMERIC/DECIMAL, FLOAT/REAL/DOUBLE, etc.).
+    - Only override types that Exasol doesn't have as first-class SQL types or
```
```suggestion
    - Only override types that Exasol doesn't have as first-class SQL type or
```

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -686,22 +697,107 @@ def visit_drop_index(self, drop, **kw):
+    - Only override types that Exasol doesn't have as first-class SQL types or
+      where SQLAlchemy may emit a name Exasol won't accept (e.g. TIME, DATETIME,
+      TEXT/LOB-ish types, UUID, binary types, ENUM).
+    - For long text, use explicit VARCHAR(n) for Alembic/reflection stability.
```
At this time, we believe but are not certain that alembic works with our dialect, so maybe it's a bit safer to leave out.
```suggestion
    - For long text, use explicit VARCHAR(n) for reflection stability.
```

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -686,22 +697,107 @@ def visit_drop_index(self, drop, **kw):
 
-    def visit_big_integer(self, type_, **kw):
-        return "DECIMAL(19)"
+    # Exasol VARCHAR max length (per docs)
```
add a link?
https://docs.exasol.com/db/latest/sql_references/data_types/datatypedetails.htm#Stringdatatypes

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -1248,3 +1389,33 @@ def fkey_rec():
+    def type_descriptor(self, typeobj):
+        """Return a DB-specific TypeEngine for a generic SA type.
+
+        We wrap DateTime columns so their Python values serialize cleanly for pyexasol.
```
Switch to official spelling.
```suggestion
        We wrap DateTime columns so their Python values serialize cleanly for PyExasol.
```

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -686,22 +697,107 @@ def visit_drop_index(self, drop, **kw):
+    # ---- helpers ----
+
+    def _varchar(self, length: int | None) -> str:
+        """Render Exasol VARCHAR(n). Clamp to max; default to max if None/0."""
```
```suggestion
        """Render Exasol VARCHAR(n). Default to max if None, 0, or the max is exceeded."""
```

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -686,22 +697,107 @@ def visit_drop_index(self, drop, **kw):
+        return self.visit_DATETIME(type_, **kw)
+
+    def visit_TIME(self, type_: sqltypes.Time, **kw: Any) -> str:
+        # If you prefer to fail fast instead, raise NotImplementedError here.
```
Not sure: Is this comment needed still?

pj-spoelders replied:
You're right. Cleaned it up.

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -769,6 +865,51 @@ def should_autocommit_text(self, statement):
         return AUTOCOMMIT_REGEXP.match(statement)
 
 
+class EXATimestamp(sqltypes.TypeDecorator):
```
If not too much effort, please move to `types.py`.

pj-spoelders replied:
Moved both types.

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -769,6 +865,51 @@ def should_autocommit_text(self, statement):
 
 
+class EXATimestamp(sqltypes.TypeDecorator):
+    """Coerce Python datetime to a JSON-serializable wire value for pyexasol.
```
```suggestion
    """Coerce Python datetime to a JSON-serializable wire value for PyExasol.
```

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -769,6 +865,51 @@ def should_autocommit_text(self, statement):
+        return value
+
+
+from sqlalchemy import exc as sa_exc
```
Move to top of the `base.py` file?

ArBridgeman commented on test/unit/exasol/dlt/test_compilers.py:
```diff
@@ -0,0 +1,146 @@
+    return base.EXATypeCompiler(base.EXADialect())
+
+
+MAX_VARCHAR = 2_000_000
```
Instead reuse ` _MAX_VARCHAR_SIZE = 2_000_000` from `base.py`?

pj-spoelders replied:
Sounds good!

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -686,22 +697,107 @@ def visit_drop_index(self, drop, **kw):
 
 
 class EXATypeCompiler(compiler.GenericTypeCompiler):
```
It'd be nice, but likely/definitely out of scope for this PR, if we had this in the user docs, relating to the types supported & transformations.

pj-spoelders committed:
Address PR feedback

pj-spoelders committed:
Address PR feedback

pj-spoelders committed:
removed failing overridden tests on returning

pj-spoelders committed:
disable uuid support because we don't have a 1-1 matching type

pj-spoelders committed:
remove unneeded import

pj-spoelders committed:
xfail datetime integration tests

pj-spoelders committed:
Fix UUID notification

pj-spoelders committed:
remove obsolete uuid test

pj-spoelders committed:
format fix

pj-spoelders committed:
xfail uuid/blob

pj-spoelders committed:
Merge branch 'master' into feature/add_dlt_support

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -1248,3 +1345,33 @@ def fkey_rec():
+
+        We wrap DateTime columns so their Python values serialize cleanly for PyExasol.
+        """
+        if isinstance(typeobj, sqltypes.DateTime):
```
Could a comment be added here about why we want this, instead of having it in colspecs? I think it'd be helpful in the future 😊

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -1248,3 +1345,33 @@ def fkey_rec():
+        return super().type_descriptor(typeobj)
+
+    # leave this for true DB-API remapping (ODBC etc.)
+    dbapi_exception_translation_map = {}
```
Since this comes from the inherited Dialect class, should this be put up at the top with the other parameters? Do we need to define it, as it looks like it has a similar definition?

<img width="1071" height="278" alt="Image" src="https://github.com/user-attachments/assets/282235ff-47f2-4c00-9003-7944b2102cd6" />

ArBridgeman commented on test/unit/exasol/dlt/test_generate_create_and_insert_statements.py:
```diff
@@ -0,0 +1,132 @@
+        """
+    )
+
+    assert _norm(insert_sql) == expected_insert
```
Would it be useful to have this also as an integration test? Then, we can verify that the converted types work as expected.

pj-spoelders replied:
I've added the integration test.

ArBridgeman commented on test/integration/sqlalchemy/test_suite.py:
```diff
@@ -94,6 +72,104 @@ class RowFetchTest(_RowFetchTest):
+        super().test_row_w_scalar_select(connection)
+
+
+DATETIME_STRING_RATIONAL = (
```
Could you add this to the `XfailRationale` class? Then use it anytime that `DATETIME_STRING_RATIONAL` and `"Websocket DBAPI currently returns DATETIME values as strings."` were used? 

Same with `UUID_BLOB_RATIONALE`.

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -686,22 +700,105 @@ def visit_drop_index(self, drop, **kw):
+      (BIGINT/INTEGER/SMALLINT, NUMERIC/DECIMAL, FLOAT/REAL/DOUBLE, etc.).
+    - Only override types that Exasol doesn't have as first-class SQL type or
+      where SQLAlchemy may emit a name Exasol won't accept (e.g. TIME, DATETIME,
+      TEXT/LOB-ish types, UUID, binary types, ENUM).
```
(out of scope) I'm not sure, but if we supported HASHTYPE https://github.com/exasol/sqlalchemy-exasol/issues/197, then I think we could support some UUIDs:
https://docs.exasol.com/db/latest/sql_references/data_types/datatypedetails.htm. Is that right?

pj-spoelders replied:
Probably! But I found the whole process to implement these from a read/write perspective too obtuse at the moment. 
This will have to wait until I get a better understanding of this or someone more familiar with the topic has time to properly implement these. 
If I have some time left this iteration or the next engineering iteration I will have another look at this if no one else tackled it by then.

ArBridgeman commented on sqlalchemy_exasol/base.py:
```diff
@@ -686,22 +700,105 @@ def visit_drop_index(self, drop, **kw):
+        return self.visit_UUID(type_, **kw)
+
+    # ---- Binary / blobs ----
+    def visit_BLOB(self, type_: sqltypes.LargeBinary, **kw: Any) -> str:
```
To the `unreleased.md`, we should:
1. Add an entry for these changes.
2. In the summary, detail what has changed:
- That BLOB, BINARY, UUID, etc. are now explicitly not supported. It'd be useful if a 
- That the exceptions raised are altered & why that's better 😄 
- List which types were altered.

For each item, a sentence or two of context would be nice.

ArBridgeman commented on test/integration/sqlalchemy/test_suite.py:
```diff
@@ -94,6 +72,104 @@ class RowFetchTest(_RowFetchTest):
+
+
+class UuidTest(_UuidTest):
+    @pytest.mark.xfail(reason=UUID_BLOB_RATIONALE, strict=True)
```
Please adapt each of these to include the specific error it will toss. It's also possible for you to just put it over the whole class like:

```python
@pytest.mark.xfail(reason=UUID_BLOB_RATIONALE, raises=CompileError, strict=True)
class UuidTest(_UuidTest):
    pass
```

ArBridgeman commented on test/integration/sqlalchemy/test_suite.py:
```diff
@@ -94,6 +72,104 @@ class RowFetchTest(_RowFetchTest):
+
+
+class DateTimeTest(_DateTimeTest):
+    @pytest.mark.xfail(
```
With the datetime change, is it possible in the code for us to convert the websocket one to the expected type? It seems like prior to this branch that this was working, so either that was an accident or sqlalchemy had a fallback to handle cases like ours. I think it could be worth looking into shortly, as this sounds like a change for customers - where the main selling part of sqlalchemy-exasol is that it's a bridge between Python & your database.

pj-spoelders committed:
Move string serialisation to websocket dialect + add condition, remove unneeded xfail blocks

pj-spoelders committed:
Ran formatting

pj-spoelders committed:
Merge branch 'master' into feature/add_dlt_support

pj-spoelders committed:
ran nox format:fix

pj-spoelders committed:
remove broken test + format

pj-spoelders committed:
PR feedback
- Removed unneeded comment
- Remove dbapi_exception_translation_map = {} statement (unneeded)

pj-spoelders committed:
Remove broken unit tests

pj-spoelders committed:
format:fix

pj-spoelders committed:
unit tests for timestamp handling

pj-spoelders committed:
move xfail rationale for UUID/BLOB tests

pj-spoelders committed:
Fix var name

pj-spoelders committed:
Add raises=compileError (PR feedback) + ran format:fix

pj-spoelders committed:
Add integration test

pj-spoelders committed:
run formatting:fix

pj-spoelders committed:
Trigger CI rerun

pj-spoelders committed:
changes to pyproject in attempt to fix black CI issue

pj-spoelders committed:
reformat files after 'poetry install'

pj-spoelders committed:
Add to  unreleased.md
